core/trino-main/src/main/java/io/trino/cache/CommonSubqueriesExtractor.java

```diff
@@ -0,0 +1,941 @@
+        return computePlanSignature(
+                commonColumnIds,
+                // Append child group by columns to signature key (if present)
+                combine(signatureKey, childAdaptation.getCommonSubplanSignature().signature().getGroupByColumns().map(columns -> "groupByColumns=" + columns).orElse("")),
+                childAdaptation.getCommonSubplanSignature().predicate(),
+                predicate,
+                projections,
+                groupByColumns,
+                childAdaptation.getCommonColumnHandles().keySet());
+    }
+
+    private PlanSignatureWithPredicate computePlanSignature(
+            Map<CacheColumnId, Symbol> commonColumnIds,
+            SignatureKey signatureKey,
+            TupleDomain<CacheColumnId> tupleDomain,
+            Expression predicate,
+            List<CacheColumnId> projections,
+            Optional<Set<CacheColumnId>> groupByColumns,
+            Set<CacheColumnId> scanColumnIds)
+    {
+        Set<CacheColumnId> projectionSet = ImmutableSet.copyOf(projections);
+        checkArgument(groupByColumns.isEmpty() || projectionSet.containsAll(groupByColumns.get()));
+
+        // Order group by columns by name
+        Optional<List<CacheColumnId>> orderedGroupByColumns = groupByColumns.map(
+                Ordering.from(comparing(CacheColumnId::toString))::immutableSortedCopy);
+        List<Type> projectionColumnsTypes = projections.stream()
+                .map(cacheColumnId -> commonColumnIds.get(cacheColumnId).type())
+                .collect(toImmutableList());
+        if (tupleDomain.isAll() && predicate.equals(TRUE)) {
+            return new PlanSignatureWithPredicate(
+                    new PlanSignature(
+                            signatureKey,
+                            orderedGroupByColumns,
+                            projections,
+                            projectionColumnsTypes),
+                    TupleDomain.all());
+        }
+
+        ExtractionResult extractionResult = DomainTranslator.getExtractionResult(
+                plannerContext,
+                session,
+                predicate);
+        // Only domains for projected columns can be part of signature predicate
+        TupleDomain<CacheColumnId> extractedTupleDomain = extractionResult.getTupleDomain()
+                .transformKeys(CanonicalSubplanExtractor::canonicalSymbolToColumnId)
+                .intersect(tupleDomain);
+        Set<CacheColumnId> retainedColumnIds = ImmutableSet.<CacheColumnId>builder()
+                // retain projected and table scan domains for per split simplification and pruning on worker node
+                .addAll(projectionSet)
+                .addAll(scanColumnIds)
+                .build();
+        TupleDomain<CacheColumnId> retainedTupleDomain = extractedTupleDomain
+                .filter((columnId, domain) -> retainedColumnIds.contains(columnId));
+        // Remaining expression and non-projected domains must be part of signature key
```

Comment:
Why so? Instead of encoding `remainingTupleDomain ` in `signatureKey`, can we make it a field of `PlanSignature`, for the case `CacheManager` is able to apply it on the cached data?
For instance, assume `CacheManager` stored `colA`, `colB`. Then, `colA` is projected with a predicate on `colA` and `colB`, the predicate on `colA` is passed as an argument to `SplitCache#loadPages` while the predicate on `colB` is encoded in the key. But `CacheManager` might be able to apply the predicate on `colA` as much as it might be able to apply the predicate on `colB`.

Reply:
When you cache data and `PlanSignature` has only `colB`, then predicate on `colA` is not meaningful for `CacheManager`.

When loading data we could pass predicate on extra columns in `loadPages` indeed.